### DISCUSSION TITLE
App Check drastically slowing down Firestore updates

### DISCUSSION THREAD

**Discussion by okcompewter** [invertase/react-native-firebase#5955] [General]
I've just added App Check to my RN app, and I immediately noticed the performance of making an update to a Firestore document is drastically impacted.

In my app.tsx:

when adding
`await firebase.appCheck().activate('ignored', true)`
an update takes about 2 seconds

when commenting this back out
`//await firebase.appCheck().activate('ignored', true)`
an update is back to taking only a tenth of a second

I didn't realize there would be such a performance hit to Firestore doc updates by enabling App Check...?

AFAIK, this slow down is only on the client side, since I have not even enabled/enforced App Check within Firebase yet.

**Comment by mikehardy** [comment]
I have noticed the AppCheck token fetch initially taking about that much time, it's a set of network round trips if I understand correctly.
Is every single document update slow, or just the first one? In my testing, it was just the initial token fetch then things were fast? But my test environment may differ - specifying exactly what your test environment is may help. At this point I'm not sure what platform you're on, whether you're using the firebase simulator, what versions, it's not possible to speculate much further

**Reply by okcompewter** [reply]
I am testing this on an Android emulated device, with a release build, against a production Firestore.

All react-native-firebase versions 14.1.0.

The slowness is actually happening on a transaction (not just a regular doc update), and happens every single time.

My transaction gets one doc, then updates 3 docs, if you'd like to try to repro.

This operation is fast without activating App Check, all other variables the same.

**Reply by mikehardy** [reply]
Very interesting! We have AppCheck enabled in our e2e rig that runs in CI, and we do hit firestore, but we hit the emulated firestore normally. It's possible to alter that by just commenting out the line that tells it to use the firestore emulator, I wonder if you altered one of our e2e tests to be non-emulator, and do the transaction like you mention (maybe add it to the issues.e2e set?) if it would reproduce. It sounds like our setups are nearly the same in CI vs your project 🤔

**Reply by mikehardy** [reply]
the other alternative is to attempt a native reproduction with firesbase-android-sdk but I think that would be very involved w/AppCheck + Firestore

**Reply by okcompewter** [reply]
I'm not sure how to try this myself (run your CI e2e rig) against an actual firebase. Any help there? Or perhaps can you just do a quick check on your end if an app with AppCheck enabled has noticeably slower transactions (2-3 seconds rather than 200-300ms)?

**Reply by mikehardy** [reply]
Run our e2e test, comment out this line https://github.com/invertase/react-native-firebase/blob/8efda506474b1757f6a0537495a1de581d372299/tests/app.js#L52

**Comment by okcompewter** [comment]
Having issue after issue just trying to get Android e2e tests to run as expected on my Windows machine.

First one is due to the `./` in front of `gradlew` in `android.emu.debug` script:
```
λ yarn tests:android:build
yarn run v1.22.17
$ cd tests && yarn detox build --configuration android.emu.debug
$ C:\Users\surfa\code\github\invertase\react-native-firebase\tests\node_modules\.bin\detox build --configuration android.emu.debug
detox[15276] INFO:  [build.js] cd android && ./gradlew assembleDebug assembleAndroidTest lintDebug -DtestBuildType=debug && cd ..'.' is not recognized as an internal or external command,
operable program or batch file.
detox[15276] WARN:  [build.js]

Important: 'detox build' is a convenience shortcut for calling your own build command, as provided in the config file.
Failures in this build command are not the responsibility of Detox. You are responsible for maintaining this command.

detox[15276] ERROR: [cli.js] Command failed: cd android && ./gradlew assembleDebug assembleAndroidTest lintDebug -DtestBuildType=debug && cd ..

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I'm ableto get around this by removing the `.\`, but then I start getting Java version errors:
```
FAILURE: Build failed with an exception.

* Where:
Build file 'C:\Users\surfa\code\github\invertase\react-native-firebase\tests\android\app\build.gradle' line: 1

* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin 'com.android.internal.application'.
   > Android Gradle plugin requires Java 11 to run. You are currently using Java 1.8.
     You can try some of the following options:
       - changing the IDE settings.
       - changing the JAVA_HOME environment variable.
       - changing `org.gradle.java.home` in `gradle.properties`.
```
I've [moved this to an issue](https://github.com/invertase/react-native-firebase/issues/5983), and hopefully someone can try to confirm the delay I'm seeing on Android with AppCheck on and doing a Firestore transaction...

**Reply by mikehardy** [reply]
:thinking: Last I checked the e2e rig was working on windows, but I haven't looked at it in a few weeks.
You're definitely going to need JDK11 though, JDK8 is effectively dead as it does not work with Android SDK 31
